Pre-fill filters when muting hashtags

shaedrich opened the pull request:
> [!WARNING]
> I haven't tested this myself (yet)

I quite like the new hashtag context menu in the Mastodon UI:
![image](https://github.com/user-attachments/assets/c258e72a-2d6d-4f6b-883b-a25cb28e1a05)

However, what bugs me a little, is, that the hashtag to mute is not prefilled, which I expected to happen from the description of the context menu item:
![image](https://github.com/user-attachments/assets/cd27313b-460c-4d96-bc6f-d79cd9bee121)

So, what this will do, is
* Add query parameters to the URL:
```diff
- /filters/new
+ /filters/new?custom_filter[keywords_attributes][0][keyword]=#dadJokes&custom_filter[keywords_attributes][0][whole_word]=1
```

This will now (if I'm not mistaken) result in the first entry of the keywords to be prefilled with "#dadJokes" and the "whole word" flag being activated for this entry

![image](https://github.com/user-attachments/assets/406a4b1a-a0cf-4625-b310-9ba5ece4f931)

shaedrich committed:
Add URL query params to "Mute hashtags" request to /filters

shaedrich committed:
Prefix hashtag text with URL-encoded hashtag character

shaedrich committed:
Pass pre-filled URL query params to keywords association build

to be pre-filled into the creation form

shaedrich committed:
Fix URL query param access

ClearlyClaire reviewed (changes requested):
Hi! Thank you for your contribution!

This is a good idea and it's going somewhere, though there are issues that need to be taken care of before this gets merged (I left inline comments).

In addition, it may be useful to have tests for such functionality. For this, we would probably be looking at system tests (those under `spec/system`).

shaedrich committed:
Merge branch 'main' into mute-prefill

shaedrich committed:
Use native escaping for hashtag URL

shaedrich commented on app/controllers/filters_controller.rb:
```diff
@@ -12,7 +12,7 @@ def index
   def new
     @filter = current_account.custom_filters.build(action: :warn)
-    @filter.keywords.build
+    @filter.keywords.build(params.dig(:custom_filter, :keywords_attributes))
```
Thanks for the explanation 👍🏻 

> That being said, this is a lot of work for just a few attributes. I wonder if we shouldn't use some bespoke parameter instead, e.g. have `/filters/new?hashtag=blah`.

And said parameter would then be handled client-side?

ClearlyClaire replied:
You would need to add handling just like you did there, e.g. `@filter.keywords.build({ 'keyword' => "##{params['hashtag']}", 'whole_word' => true }) if params['hashtag'].present?`

shaedrich committed:
Use validated resource_params instead of plain params

… as well as properly fill :keywords_attributes as collection rather than single model

Co-authored-by: Claire <claire.github-309c@sitedethib.com>

See: https://github.com/mastodon/mastodon/pull/35133#discussion_r2163366096

Co-authored-by: Claire <claire.github-309c@sitedethib.com>

shaedrich committed:
Merge branch 'main' into mute-prefill

shaedrich committed:
fix const assignment

shaedrich committed:
Explicitly type variable

See https://github.com/mastodon/mastodon/actions/runs/15860724453

shaedrich committed:
Pass hashtagUrl variable to useMemo() as dependency

See https://github.com/mastodon/mastodon/actions/runs/15860724453

shaedrich committed:
Merge branch 'main' into mute-prefill

shaedrich committed:
Revert "Pass hashtagUrl variable to useMemo() as dependency"

This reverts commit 59abde12ee45b507cfdbd4b997d1f61817e3f9a5.

shaedrich committed:
Merge branch 'main' into mute-prefill

shaedrich committed:
Fix variable types

shaedrich committed:
fix: :bug: Add hashtagUrl to target params

so that the variable is avialable outside of useEffect()

shaedrich committed:
Linter

shaedrich committed:
'hashtagUrl' is possibly 'undefined'

shaedrich committed:
Pass hashtagUrl variable to useMemo() as dependency

Redo 59abde1

cheeaun commented:
Question, would this work for adding to existing filters?

mjankowski commented:
I see why you picked the `keywords_attributes[0]` approach here to make use of existing nested attributes ... but I wonder about the awkwardness of that from an API perspective? It's sort of borrowing an implementation detail of that form and calling in from elsewhere.

Given we don't currently support any data population at all, would a better first pass here be to just allow like `/filters/new?value=xxx` (or whatever naming) and then handle the data shuffling in the controller, making sure we continue to support the current feature when nothing provided? I assume that'd be a little more straightforward on the JS side.

shaedrich commented:
> Question, would this work for adding to existing filters?

@cheeaun Nope. But how could that look like?

shaedrich commented:
> I see why you picked the `keywords_attributes[0]` approach here to make use of existing nested attributes ... but I wonder about the awkwardness of that from an API perspective? It's sort of borrowing an implementation detail of that form and calling in from elsewhere.
> 
> Given we don't currently support any data population at all, would a better first pass here be to just allow like `/filters/new?value=xxx` (or whatever naming) and then handle the data shuffling in the controller, making sure we continue to support the current feature when nothing provided? I assume that'd be a little more straightforward on the JS side.

@mjankowski Let's assume, we call it `/filters/new?keyword=xxx` (implementation detail, I know, but it would be confusing otherwise, I think), this would allow us to prefill the `keyword` server-side, but we can't just set the `whole_word` checkbox by default as this would also be confusing. As this won't work, we could try a more expressive route: `/filters/new/tag/xxx` This works for hashtags, because they don't contain any special characters (right?). It being hashtag-specific might give us one more single-purpose route to maintain but we would gain a way to explicitly pre-populate the `whole_word` checkbox as well—what do you think? Of course, other, similar routes might not work this way, but `/filters/new/tag?tag=xxx` feels unnecessarily redundant to me 🤔

shaedrich commented:
Sorry for abandoning this PR. I vaguely remember that the basic implementation worked but there was one small thing (reactivity perhaps?) that I couldn't make work. Out of frustration, I gave up. But I'm still interested in getting this solved, just that I might not be able to solve it entirely on my own.
